Add PyVista for 3D plotting in Python

banesullivan opened the pull request:
### Todo

- [x] Make PyVista an optional dependency
- [x] Address #30 before merging. 

-----

This PR uses [PyVista](http://docs.pyvista.org) for creating VTK datasets in Python for immediate plotting. The `pyevtk`  export code is still present to allow users to save VTK files in any Python environment. PyVista is a heavy dependency so these additions are implemented as optional features.

When complete, this will need to use "Squash and Merge"

-----

PyVista allows you to create VTK data objects in memory really easily instead of saving them out to static files. *Why might you want VTK/PyVista meshes in Python?* So you can skip saving VTK files and visualize the datasets right in Python! Check out this example:

```py
import pyvista as pv
from gstools import SRF, Gaussian
import matplotlib.pyplot as pt

# structured field with a size 100x100x100 and a grid-size of 1x1x1
x = y = z = range(100)
model = Gaussian(dim=3, var=0.6, len_scale=20)
srf = SRF(model)
srf((x, y, z), mesh_type='structured')

# Create the VTK mesh
mesh = srf.to_pyvista()

# Quickly plot in 3D:
mesh.plot()
```

![download](https://user-images.githubusercontent.com/22067021/65274009-a1e1c200-dadf-11e9-8744-5694c96b1526.png)

```py
# Use PyVista for some interactive plotting
p = pv.Plotter(notebook=False)
p.add_mesh_threshold(mesh)
p.show_grid()
p.show()
```
![2019-09-19 13 10 31](https://user-images.githubusercontent.com/22067021/65274017-a60ddf80-dadf-11e9-9fe2-a27e1197c4a2.gif)



### Notes

I added a new method: `to_pyvista()`.

banesullivan commented:
These CI scripts are a mess... AppVeyor is all mixed up between Python 2 and 3 and needs Anaconda to manage everything. Meanwhile, Travis is being prompted for Twine upload passwords and it's not properly setting up VTK on Linux 🤦‍♂

banesullivan commented:
I think the docker container on Travis isn't properly reflecting the dependencies I'm trying to install:

```
sudo apt update && sudo apt install python-qt4 libgl1-mesa-glx
```

banesullivan commented:
AppVeyor build seems hung on the twine upload?

LSchueler commented:
Wow, thank you so much for the effort you already put into this project!

I haven't looked into the details yet, but your plotting routines look very intriguing indeed.
Depending on how heavy the dependencies actually become, we might just put the imports into the according functions, in order to only import them when needed.

MuellerSeb commented:
@banesullivan : Thanks for this astonishing PR!
The pictures speak for themselves.
I'd love to have such an interface to pyvista. But I would suggest another way:
- as you said, pyevtk is lightweight and therefore I would keep it as the standard way of exporting
  (at least for the 1.1.x versions of GSTools)
- make pyvista an optional dependency, with the following line in the setup.py:
  ``extras_require={"show": ["matplotlib", "pyvista"]}``
  (this also adds matplotlib for plotting)
- as @LSchueler proposed, import pyvista within the called method (maybe check for ImportError)
- keep the CI the way it is, as for the 1.1.x versions we want to still provide support for 2.7 and >=3.4
  (we have in mind to drop py2 support from 2020 on as you commented in https://github.com/GeoStat-Framework/GSTools/issues/30)
- in field/base.py: 
  - I would just add the ``to_pyvista`` method, that returns the pyvista object (then users can continue with the pyvista infrastructure)
  - keep the ``export_vtk`` routine
- in tools/export.py:
  - just add a ``to_pyvista`` function (taking a mesh_type argument)

I hope you are OK with this. We don't want to drop so much supported version for a visualization routine. At least at the moment.

What do you think?

Thanks again for your work and the nice pyvista package!

Cheers, Sebastian

banesullivan commented:
Hi @MuellerSeb, thanks for all the feedback! I'll get back to this PR soon and clean it up so PyVista is simply added on top as an extra and rebase all these commits to leave the CI integration untouched.

Overall, I think what you propose is perfect

banesullivan commented:
@LSchueler and @MuellerSeb, I just rebased all of the commits and made the desired changes! I also updated the description in the first post (^^^), take a look and let me know what you think!

Notes:

- `pyevtk` is still used as the defauly method for exporting VTK datasets so all Python versions are supported.
- PyVista is now listed as an extra dependency with matplotlib in the `setup.py` for the `plotting` option.
- If a user calls `to_pyvista` and PyVista is unavailable, they are prompted with a friendly `ImportError`
- No changes to the CI integration or wheel building

banesullivan commented:
Also, I added tests for both `pyevtk` and PyVista. The PyVista tests will not run unless PyVista is installed... so the CIs will not run those tests, but I could set that up for Python version greater than 3.4 if desired.

MuellerSeb commented on requirements.txt:
```diff
@@ -0,0 +1,10 @@
+six
+matplotlib
+pytest
+pytest-cov
```
Please remove pyvista, matplotlib, pytest and pytest-cov from this list.

MuellerSeb replied:
Because of this, the py34 build fails and we don't need to install pyvista and matplotlib on travis.

MuellerSeb commented:
Also don't forget to add yourself to the AUTHORS.md under Contributors! ;-)

MuellerSeb commented:
> AppVeyor build seems hung on the twine upload?

You don't have access to our PyPI password, that is why both, travis and appveyor, are hanging on twine and the CI fails. So this is no problem.

MuellerSeb commented on requirements.txt:
```diff
@@ -3,8 +3,5 @@ scipy>=1.1.0
-pyvista
 six
-matplotlib
 pytest
```
Please also remove "pytest"

banesullivan commented:
Looking good to me! I look forward to using this package - I've been looking for a good kriging toolbox and I think I've found it! Having the ties to PyVista might bring in a few other users if not just myself!

LSchueler commented on README.md:
```diff
@@ -104,7 +104,7 @@ srf.plot()
 </p>
 
-A similar example but for a three dimensional field is exported to a [VTK](https://vtk.org/) file, which can be visualized with [ParaView](https://www.paraview.org/).
+A similar example but for a three dimensional field is exported to a [VTK](https://vtk.org/) file, which can be visualized with [ParaView](https://www.paraview.org/) or [PyVista](http://docs.pyvista.org) in Python:
```
Please change the URL to https://docs.pyvista.org. We don't want to encourage unsafe connections ;-)

LSchueler commented on gstools/field/base.py:
```diff
@@ -147,16 +147,15 @@ def mesh(
         ----------
-        filename : :class:`str`
-            Filename of the file to be saved, including the path. Note that an
-            ending (.vtr or .vtu) will be added to the name.
```
I guess you accidentally deleted the parameter description of "filename". Please undo the deletion and add a description for the different handling of None / not None.

LSchueler commented on gstools/tools/export.py:
```diff
@@ -53,27 +54,56 @@ def vtk_export_structured(filename, pos, fields):  # pragma: no cover
+    fields : :class:`dict` or :class:`numpy.ndarray`
+        Structured fields to be saved.
+        Either a single numpy array as returned by SRF,
+        or a dictionary of fields with theirs names as keys.
```
Please provide a description of what is being returned.

LSchueler commented on gstools/tools/export.py:
```diff
@@ -53,27 +54,56 @@ def vtk_export_structured(filename, pos, fields):  # pragma: no cover
-        Unstructured fields to be saved.
+        Structured fields to be saved.
         Either a single numpy array as returned by SRF,
         or a dictionary of fields with theirs names as keys.
```
Please provide a description of what is being returned.

banesullivan replied:
Nothing is returned here. it is just good practice to return the call to the function that this function wraps.

LSchueler commented on gstools/tools/export.py:
```diff
@@ -90,9 +120,75 @@ def vtk_export_unstructured(filename, pos, fields):  # pragma: no cover
+    fields : :class:`dict` or :class:`numpy.ndarray`
+        Unstructured fields to be saved.
+        Either a single numpy array as returned by SRF,
+        or a dictionary of fields with theirs names as keys.
```
Please provide a description of what is being returned.

LSchueler commented on gstools/tools/export.py:
```diff
@@ -90,9 +120,75 @@ def vtk_export_unstructured(filename, pos, fields):  # pragma: no cover
+    fields : :class:`dict` or :class:`numpy.ndarray`
+        Unstructured fields to be saved.
+        Either a single numpy array as returned by SRF,
+        or a dictionary of fields with theirs names as keys.
```
Please provide a description of what is being returned.

banesullivan replied:
Same no return type. Simply good practice

LSchueler commented on gstools/tools/export.py:
```diff
@@ -90,9 +120,75 @@ def vtk_export_unstructured(filename, pos, fields):  # pragma: no cover
+        Either a single numpy array as returned by SRF,
+        or a dictionary of fields with theirs names as keys.
+    mesh_type : :class:`str`, optional
+        'structured' / 'unstructured'. Default: structured
```
Please provide a description of what is being returned.
